pennylane/gradients/parameter_shift.py

```diff
@@ -0,0 +1,450 @@
+of a qubit-based quantum tape.
+"""
+# pylint: disable=protected-access,too-many-arguments
+import numpy as np
+
+import pennylane as qml
+
+from .finite_difference import finite_diff, generate_shifted_tapes
+
+
+def _get_operation_recipe(tape, t_idx, shift=np.pi / 2):
+    """Utility function to return the parameter-shift rule
+    of the operation corresponding to trainable parameter
+    t_idx on tape.
+
+    If the corresponding operation has grad_recipe=None, then
+    the default two-term parameter-shift rule is assumed.
+    """
+    p_idx = list(tape.trainable_params)[t_idx]
+    op = tape._par_info[p_idx]["op"]
+    op_p_idx = tape._par_info[p_idx]["p_idx"]
+    return op.get_parameter_shift(op_p_idx, shift=shift)
+
+
+def _process_gradient_recipe(gr, tol=1e-10):
+    """Utility function to process gradient recipes."""
+    gr = np.array(gr).T
+    # remove all small coefficients, shifts, and multipliers
+    gr[np.abs(gr) < tol] = 0
+    # remove columns where the coefficients are 0
+    gr = gr[:, ~(gr[0] == 0)]
+    # sort columns according to abs(shift)
+    return gr[:, np.argsort(np.abs(gr)[-1])]
+
+
+def _gradient_analysis(tape, use_graph=True):
+    """Update the parameter information dictionary  of the tape with
+    gradient information of each parameter."""
+
+    if getattr(tape, "_gradient_fn", None) is param_shift:
+        # gradient analysis has already been performed on this tape
+        return
+
+    tape._gradient_fn = param_shift
+
+    for idx, info in tape._par_info.items():
+
+        if idx not in tape.trainable_params:
+            info["grad_method"] = None
+        else:
+            op = tape._par_info[idx]["op"]
+
+            if op.grad_method == "F":
+                info["grad_method"] = "F"
+            else:
+                info["grad_method"] = tape._grad_method(
+                    idx, use_graph=use_graph, default_method="A"
+                )
+
+
+def expval_param_shift(tape, argnum=None, shift=np.pi / 2, gradient_recipes=None, f0=None):
+    r"""Generate the parameter-shift tapes and postprocessing methods required
+    to compute the gradient of an gate parameter with respect to an
+    expectation value.
+
+    Args:
+        tape (.QuantumTape): quantum tape to differentiate
+        argnum (int or list[int] or None): Trainable parameter indices to differentiate
+            with respect to. If not provided, the derivative with respect to all
+            trainable indices are returned.
+        shift (float): The shift value to use for the two-term parameter-shift formula.
+            Only valid if the operation in question supports the two-term parameter-shift
+            rule (that is, it has two distinct eigenvalues) and ``gradient_recipe``
+            is ``None``.
+        gradient_recipes (tuple(list[list[float]] or None)): List of gradient recipes
+            for the parameter-shift method. One gradient recipe must be provided

```

Comment:
It is not being checked, _but_ the `for` loop will fail if this is not the case.

Due to performance, I try to avoid validating everything - I hope that the docstrings will suffice is most cases! Let me know what you think